src/v/security/role_store.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "absl/container/node_hash_map.h"
+#include "security/role.h"
+#include "security/types.h"
+
+#include <seastar/util/variant_utils.hh>
+
+#include <ranges>
+
+namespace security {
+
+/* container for Roles */
+class role_store {
+    using role_types = std::variant<role>;
+    using container_type = absl::node_hash_map<role_name, role_types>;
+
+public:
+    role_store() noexcept = default;
+    role_store(const role_store&) = delete;
+    role_store& operator=(const role_store&) = delete;
+    role_store(role_store&&) noexcept = default;
+    role_store& operator=(role_store&&) noexcept = default;
+    ~role_store() noexcept = default;
+
+    template<typename T>
+    void put(const role_name& name, T&& role) {
+        _roles.insert_or_assign(name, std::forward<T>(role));
+    }
+
+    template<typename T>
+    auto get(const role_name& name) -> std::optional<T> const {
+        if (auto it = _roles.find(name); it != _roles.end()) {
+            return std::get<T>(it->second);
+        }
+        return std::nullopt;
+    }
+
+    bool remove(const role_name& name) { return _roles.erase(name) > 0; }
+
+    bool contains(const role_name& name) const { return _roles.contains(name); }
+
+    auto range(auto pred) {
+        return _roles | std::views::filter(std::move(pred));
```

Comment:
Pretty sure you can also use forwarding reference here with `auto&&`?

Reply:
probably so, yeah 👍

Reply:
done